Add a splash screen to RuneLite

jplsek opened the pull request:
## About
Although the client is still fast when starting up, "nothing" happens for several seconds. Adding a splash screen at least lets the user know that "something" is happening, and not try to open the client multiple times. This is more noticeable when the api is acting slow or when users have a slow connection. The current implementation does not seem to affect launch time.

## API
- All public methods are non-blocking.
- There is an argument to disable the splash screen.

I'm not sure how the API should act. Right now, it's part of `init`. Should the `setMessage` calls be inside the individual methods?

## UI
Updated:
![updated4](https://user-images.githubusercontent.com/2388657/40864545-1e66527e-65ec-11e8-9f73-db571bb30fe4.gif)

Before:
<details>

Here is a full video of the splash screen (the inconsistent "RuneScape" capitals have been fixed):
![normal](https://user-images.githubusercontent.com/2388657/40737742-8a670748-6439-11e8-996e-f88740dcd4cd.gif)

Slow connection simulation (1000ms delay):
![slow](https://user-images.githubusercontent.com/2388657/40737757-96746152-6439-11e8-8024-c3bfc9a5c69d.gif)
</details>

Feedback welcome. Thanks!

deathbeam commented:
You could also instantly show the main frame after loading the configuration, then there is very small delay, because most of it is mostly just plugins starting up and all the services.

jplsek commented:
Hmm, that's a good point. When I tried showing the UI early, the frame/applet will resize when I place the `clientUI.show()` before `sessionManager.loadSession()`. Not sure why. (I have not looked into it yet.)

psikoi commented:
Are you set on that layout/presentation? I like the idea of a splash/loading screen but I think it could be better presented. I can help if you dont mind

jplsek commented:
Sure thing! Concepts could be posted here or/and I can give you access to modify the branch.

psikoi commented:
Based on that first example you gave, these are my thoughts:

- The long description/title seems unnecessary
- The logo is too big I feel, this is a logo that reads well at small sizes so there's no reason to have it that big
- The positioning on the version and status message is a bit weird
- The spinning wheel usually serves as a loading indicator for small periods of time, loading bar usually for longer periods of time, having both makes no sense, I think the bar is a better indicator to have.

Okay so here's my improvement idea:

![](https://i.gyazo.com/63baf6735808ee83ef4d53f9853eeefe.png)

Abextm commented:
- We should probably use [the icon without the background](https://user-images.githubusercontent.com/12366911/40821976-0ad5d042-6527-11e8-8cd7-c0fb8a75dd8f.png)
 - The splash should probably show before guice starts injecting, as a large part of startup time is injection and it is very easy to make changes that could affect it negativly
 - We should probably show the same/similar UI in the launcher, replacing the current download bar.

psikoi commented:
Oh I wasn't aware there was a version without the background, give me a second.

psikoi commented:
With transparent logo:

![](https://i.gyazo.com/726fcb4d60f254bb1bf302c1ab3fedf0.png)

jplsek commented:
Alright, updated to @psikoi's design!

I changed the L&F of the progress bar through the UI Manager (in case of future usage of JProgressBar). If this is undesired, I can set the colors directly to the progress bar.

The colors are slightly different, just to follow the current scheme and using Java's helper methods. There's a similar situation to the vertical spacing of components.

I was unable to make the font size smaller for the version and messages without the font breaking.

I looked into starting before guice, but for me, guice injection was so fast that the frame didn't show up until it already finished the injection process. This can still be done if requested.

![updated](https://user-images.githubusercontent.com/2388657/40856733-c971577a-65d0-11e8-81f1-f1ea051b91d7.gif)

I would prefer to have a spinner somewhere so users don't think the frame is frozen. I'm not sure how to cleanly integrate that though.

psikoi commented:
Did you try using the small font?

jplsek commented:
Ah, didn't know about that! Will update soon.

psikoi commented:
Can you add padding to the status message? That would wrap the text when it's too long, making it not so close to the edges

jplsek commented:
Since the statuses are using JLabels, they can't be multi-line unless html tags are used. I'd like to avoid multi-line statuses for style and readability. The layout will also have to be modified to be less dynamic, since adding lines will automatically change the vertical padding to center everything. Also with the smaller font, and the current longest text, there is decent horizontal padding.

psikoi commented:
Could the name and version be moved up q bit? Sorry ill stop after this i promise

jplsek commented:
No problem! Updated the main post.

deathbeam commented:
Needs rebase. Also, this should match the launcher loading, having 2 different loading screens is weird. Another option is to update also launcher loading, but that seems less feasible.

Abextm commented:
It would be nice if the launcher would spawn the window, then the client could update the launcher's progress over a local socket, or spawn it's own window if it can't find the launcher.

deathbeam commented:
With ReflectionLauncher we would not even need local socket, but that would not work for JVMLauncher unfortunately. But otherwise yeah, that would feel pretty seamless.

jplsek commented:
Rebased.

Making the launcher match L&F of the client requires duplicating code (SwingUtil, FontManager, ColorScheme, SubstanceRuneLiteLookAndFeel, ObsidianTheme, probably more) and bring in the Substance theme to the launcher, which makes it slightly more annoying to maintain.

But if that's fine, then I could try to look into that.

After that, I could look into making a socket and essentially remove the RuneLiteSplashScreen from the client as mentioned before.

But, if the client is sending commands to the launcher, that means the client would close the launcher. What happens if the client never closes the launcher (like the client crashing)? Should there be a timeout to just kill the launcher after a certain amount of time? I think this is making things too complicated/hackish.

Adam- commented:
This needs rebase because plugin starting was moved after the UI shows to make the client appear to start faster.

I would have the loading stages be:

1) pre guice ("Building client"?)
2) pre config+session load
3) pre plugin loading
4) pre session loading

After this the UI is visible, and then the plugin startup happens.

Adam- commented on runelite-client/src/main/java/net/runelite/client/RuneLiteSplashScreen.java:
```diff
@@ -0,0 +1,218 @@
+ */
+@Slf4j
+@Singleton
+public class RuneLiteSplashScreen
```
This should probably be under `net.runelite.client.ui`

james-munson commented:
@jplsek ah i was using the older branch, pulling this one to see if it works

james-munson commented:
yep that fixed it awesome

deathbeam commented:
Well my comment still applies, this "splash screen" is very different from launcher and so either launcher or this needs to be changed so the 2 would match
